tracer/src/Datadog.Trace/FeatureFlags/FeatureFlagsEvaluator.cs

```diff
@@ -0,0 +1,704 @@
+                    }
+
+                    if (allocation.Splits is { Count: > 0 })
+                    {
+                        foreach (var split in allocation.Splits)
+                        {
+                            if (split.Shards is { Count: 0 })
+                            {
+                                return ResolveVariant(flagKey, resultType, defaultValue, flag, split.VariationKey ?? string.Empty, allocation, context);
+                            }
+
+                            var allShardsMatch = true;
+                            foreach (var shard in split.Shards!)
+                            {
+                                if (!MatchesShard(shard, targetingKey))
+                                {
+                                    allShardsMatch = false;
+                                    break;
+                                }
+                            }
+
+                            if (allShardsMatch)
+                            {
+                                return ResolveVariant(flagKey, resultType, defaultValue, flag, split.VariationKey ?? string.Empty, allocation, context);
+                            }
+                        }
+                    }
+                }
+
+                // No allocation / split matched – use default
+                return new Evaluation(
+                    flagKey,
+                    defaultValue,
+                    EvaluationReason.DEFAULT);
+            }
+            catch (FormatException ex)
+            {
+                Log.Debug(ex, "Evaluation failed for key {Key}", flagKey);
+                return new Evaluation(
+                    flagKey,
+                    defaultValue,
+                    EvaluationReason.ERROR,
+                    error: "TYPE_MISMATCH",
+                    metadata: new Dictionary<string, string>
+                    {
+                        ["errorCode"] = "TYPE_MISMATCH"
+                    });
+            }
+            catch (Exception ex)
+            {
+                Log.Debug(ex, "Evaluation failed for key {Key}", flagKey);
+                return new Evaluation(
+                    flagKey,
+                    defaultValue,
+                    EvaluationReason.ERROR,
+                    error: ex.Message,
+                    metadata: new Dictionary<string, string>
+                    {
+                        ["errorCode"] = "GENERAL",
+                        ["message"] = ex.Message
+                    });
+            }
+        }
+
+        // --------- helpers ----------
+
+        private static bool IsAllocationActive(Allocation allocation, DateTime now)
+        {
+            var startDate = ParseDate(allocation.StartAt);
+            if (startDate.HasValue && now < startDate.Value)
+            {
+                return false;
+            }
+
+            var endDate = ParseDate(allocation.EndAt);
+            if (endDate.HasValue && now > endDate.Value)
```

Comment:
nitpick: I believe end date shall be exclusive
```suggestion
            if (endDate.HasValue && now >= endDate.Value)
```